From 5f62f5e4ac22986130aeb30915b872a92b1f0d25 Mon Sep 17 00:00:00 2001
From: Matus Kral <matuskral@me.com>
Date: Wed, 18 Mar 2015 13:03:59 +0100
Subject: IMX6: threaded PA change reporter to avoid blocking process()

    - when transmitting PA, CECHandler transmits PA with bSuppressWait
set to NO. Waiting here will block main Process() which should
receive NACK/ACK (but it can't because it is it who called callback
which is triggering PA resending)
---
 src/libcec/adapter/IMX/IMXCECAdapterCommunication.cpp | 32 ++++++++++++++++++----
 src/libcec/adapter/IMX/IMXCECAdapterCommunication.h   | 13 +++++++++
 2 files changed, 40 insertions(+), 5 deletions(-)

diff --git a/src/libcec/adapter/IMX/IMXCECAdapterCommunication.cpp b/src/libcec/adapter/IMX/IMXCECAdapterCommunication.cpp
index f1cb033..b2da80b 100644
--- a/src/libcec/adapter/IMX/IMXCECAdapterCommunication.cpp
+++ b/src/libcec/adapter/IMX/IMXCECAdapterCommunication.cpp
@@ -34,6 +34,7 @@
 #include "lib/LibCEC.h"
 #include "lib/platform/sockets/cdevsocket.h"
 #include "lib/platform/util/StdString.h"
+#include "platform/util/util.h"
 
 using namespace std;
 using namespace CEC;
@@ -54,7 +55,8 @@ using namespace PLATFORM;
 #define CEC_MSG_FAIL_DATA_NOT_ACK       0x86	/*Message transmisson failed: Databyte not acknowledged*/
 
 CIMXCECAdapterCommunication::CIMXCECAdapterCommunication(IAdapterCommunicationCallback *callback) :
-    IAdapterCommunication(callback)
+    IAdapterCommunication(callback),
+    m_PAReporter(NULL)
 {
   CLockObject lock(m_mutex);
 
@@ -63,11 +65,13 @@ CIMXCECAdapterCommunication::CIMXCECAdapterCommunication(IAdapterCommunicationCa
   m_bLogicalAddressRegistered = false;
   m_bInitialised = false;
   m_dev = new CCDevSocket(CEC_IMX_PATH);
+  m_physicalAddress = -1;
 }
 
 CIMXCECAdapterCommunication::~CIMXCECAdapterCommunication(void)
 {
   Close();
+  DELETE_AND_NULL(m_PAReporter);
   delete m_dev;
   m_dev = 0;
 }
@@ -334,23 +338,41 @@ void *CIMXCECAdapterCommunication::Process(void)
         {
             /* HDMI Hotplug event - disconnect */
         }
-        else if (event.event_type == MESSAGE_TYPE_CONNECTED)
+        else if (event.event_type == MESSAGE_TYPE_CONNECTED && m_physicalAddress != 0xffff)
         {
             /* HDMI Hotplug event - connect */
             uint16_t oldAddress = m_physicalAddress;
 
-            if (oldAddress != GetPhysicalAddress())
-              m_callback->HandlePhysicalAddressChanged(m_physicalAddress);
+            if (oldAddress != GetPhysicalAddress()) {
+              if (m_PAReporter)
+                while (m_PAReporter->IsRunning()) Sleep(5);
+              delete m_PAReporter;
+
+              m_PAReporter = new CCECPAChangedReporter(m_callback, m_physicalAddress);
+              m_PAReporter->CreateThread();
+            }
 #ifdef CEC_DEBUGGING
             LIB_CEC->AddLog(CEC_LOG_DEBUG, "%s: plugin event received", __func__);
 #endif
         }
         else
             LIB_CEC->AddLog(CEC_LOG_WARNING, "%s: unhandled response received %d!", __func__, event.event_type);
-    }       
+    }
   }
 
   return 0;
 }
 
+CCECPAChangedReporter::CCECPAChangedReporter(IAdapterCommunicationCallback *callback, uint16_t newPA) :
+    m_callback(callback),
+    m_newPA(newPA)
+{
+}
+
+void* CCECPAChangedReporter::Process(void)
+{
+  m_callback->HandlePhysicalAddressChanged(m_newPA);
+  return NULL;
+}
+
 #endif	// HAVE_IMX_API
diff --git a/src/libcec/adapter/IMX/IMXCECAdapterCommunication.h b/src/libcec/adapter/IMX/IMXCECAdapterCommunication.h
index 8ba6c43..8411582 100644
--- a/src/libcec/adapter/IMX/IMXCECAdapterCommunication.h
+++ b/src/libcec/adapter/IMX/IMXCECAdapterCommunication.h
@@ -54,6 +54,17 @@ namespace CEC
 {
   class CIMXCECAdapterMessageQueueEntry;
 
+  class CCECPAChangedReporter : public PLATFORM::CThread
+  {
+  public:
+    CCECPAChangedReporter(IAdapterCommunicationCallback *callback, uint16_t newPA);
+    void* Process(void);
+
+  private:
+    IAdapterCommunicationCallback *m_callback;
+    uint16_t                       m_newPA;
+  };
+
   class CIMXCECAdapterCommunication : public IAdapterCommunication, public PLATFORM::CThread
   {
   public:
@@ -119,6 +130,8 @@ namespace CEC
 
     bool                        m_bLogicalAddressRegistered;
     bool                        m_bInitialised;
+
+    CCECPAChangedReporter       *m_PAReporter;
   };
   
 };
-- 
1.9.1

